File: examples/anthropic-managed-agents/python/anthropic_managed_agents_e2b/app_webhook_server.py
from __future__ import annotations

import asyncio
import hmac
import logging
from threading import Lock

import anthropic
from fastapi import FastAPI, HTTPException, Request, Response

from anthropic_managed_agents_e2b.app_sandbox_store import JsonSandboxStore
from anthropic_managed_agents_e2b.sandbox_worker import ensure_worker_sandbox
from anthropic_managed_agents_e2b.settings import (
    DEFAULT_APP_SANDBOX_TIMEOUT_SECONDS,
    DEFAULT_LOG_LEVEL,
    DEFAULT_SANDBOX_TIMEOUT_SECONDS,
    DEFAULT_TEMPLATE_NAME,
    DEFAULT_WORKER_MAX_IDLE_SECONDS,
    MAX_WEBHOOK_BODY_BYTES,
    Settings,
    load_settings,
)

app = FastAPI()
store = JsonSandboxStore()
worker_locks: dict[tuple[str, str, str], Lock] = {}
worker_locks_lock = Lock()
queue_drains: dict[str, asyncio.Task[None]] = {}
logger = logging.getLogger(__name__)
ROUTING_SCOPES = {"session", "agent", "environment"}


def _settings() -> Settings:
    return load_settings()


def _webhook_client(settings: Settings) -> anthropic.Anthropic:
    return anthropic.Anthropic(api_key=settings.require_anthropic_api_key())


def _async_webhook_client(settings: Settings) -> anthropic.AsyncAnthropic:
    return anthropic.AsyncAnthropic(api_key=settings.require_anthropic_api_key())


def _routing_scope(settings: Settings) -> str:
    scope = settings.app_sandbox_routing_scope or "session"
    if scope not in ROUTING_SCOPES:
        raise RuntimeError("APP_SANDBOX_ROUTING_SCOPE must be session, agent, or environment")
    return scope


def _routing_target(settings: Settings, session_id: str) -> tuple[str, str, str]:
    configured_environment_id = settings.require_anthropic_environment_id()
    scope = _routing_scope(settings)
    if scope == "session":
        return configured_environment_id, scope, session_id
    if scope == "environment":
        return configured_environment_id, scope, configured_environment_id

    session = _webhook_client(settings).beta.sessions.retrieve(session_id)
    if session.environment_id != configured_environment_id:
        raise RuntimeError(
            f"session {session_id} belongs to {session.environment_id}, "
            f"but this worker is configured for {configured_environment_id}"
        )
    return session.environment_id, scope, session.agent.id


def ensure_worker_for_work(settings: Settings, work: object):
    data = getattr(work, "data", None)
    if getattr(data, "type", None) != "session":
        return None
    session_id = getattr(data, "id", None)
    if not session_id:
        return None
    work_id = getattr(work, "id", None)
    if not work_id:
        raise RuntimeError("claimed work item does not include id")
    session_id = str(session_id)
    environment_id, routing_scope, routing_id = _routing_target(settings, session_id)
    with worker_locks_lock:
        worker_lock = worker_locks.setdefault((environment_id, routing_scope, routing_id), Lock())

    with worker_lock:
        return _ensure_worker_for_target(
            settings,
            environment_id,
            routing_scope,
            routing_id,
            work_id=str(work_id),
            session_id=session_id,
        )


def _ensure_worker_for_target(
    settings: Settings,
    environment_id: str,
    routing_scope: str,
    routing_id: str,
    work_id: str,
    session_id: str,
):
    assignment = store.get(
        environment_id=environment_id, routing_scope=routing_scope, routing_id=routing_id
    )
    sandbox = ensure_worker_sandbox(
        settings,
        template_name=DEFAULT_TEMPLATE_NAME,
        timeout_seconds=(
            DEFAULT_APP_SANDBOX_TIMEOUT_SECONDS
            if routing_scope == "session"
            else DEFAULT_SANDBOX_TIMEOUT_SECONDS
        ),
        worker_max_idle_seconds=DEFAULT_WORKER_MAX_IDLE_SECONDS,
        log_level=DEFAULT_LOG_LEVEL,
        work_id=work_id,
        session_id=session_id,
        sandbox_id=assignment.sandbox_id if assignment else None,
    )
    store.upsert(
        environment_id=environment_id,
        routing_scope=routing_scope,
        routing_id=routing_id,
        session_id=session_id,
        sandbox_id=sandbox.sandbox_id,
    )
    return sandbox


async def drain_work_queue(settings: Settings) -> None:
    environment_id = settings.require_anthropic_environment_id()
    async with _async_webhook_client(settings) as client:
        async for work in client.beta.environments.work.poller(
            environment_id=environment_id,
            environment_key=settings.require_anthropic_environment_key(),
            drain=True,
            auto_stop=False,
        ):
            await asyncio.to_thread(ensure_worker_for_work, settings, work)


def _log_background_queue_result(environment_id: str, task: asyncio.Task[None]) -> None:
    queue_drains.pop(environment_id, None)
    try:
        task.result()
    except Exception:
        logger.exception("failed to drain work queue")
        return

    logger.info("drained work queue for %s", environment_id)


def start_queue_drain(settings: Settings) -> None:
    environment_id = settings.require_anthropic_environment_id()
    if environment_id in queue_drains:
        return

    task = asyncio.create_task(drain_work_queue(settings))
    queue_drains[environment_id] = task
    task.add_done_callback(lambda done: _log_background_queue_result(environment_id, done))


async def _read_limited_body(request: Request, max_bytes: int) -> str:
    content_length = request.headers.get("content-length")
    if content_length and int(content_length) > max_bytes:
        raise ValueError("request body too large")

    chunks: list[bytes] = []
    size = 0
    async for chunk in request.stream():
        size += len(chunk)
        if size > max_bytes:
            raise ValueError("request body too large")
        chunks.append(chunk)
    return b"".join(chunks).decode()


def _has_admin_access(request: Request, settings: Settings) -> bool:
    expected = settings.app_webhook_admin_token
    authorization = request.headers.get("authorization", "")
    prefix = "Bearer "
    actual = authorization[len(prefix) :] if authorization.startswith(prefix) else ""
    return bool(expected and actual and hmac.compare_digest(expected, actual))


@app.get("/health")
def health() -> dict[str, bool]:
    return {"ok": True}


@app.get("/sandboxes")
def sandboxes(request: Request) -> dict[str, list[dict[str, str]]]:
    if not _has_admin_access(request, _settings()):
        raise HTTPException(status_code=401, detail="unauthorized")

    return {"sandboxes": [assignment.__dict__ for assignment in store.list()]}


@app.post("/webhook")
async def webhook(request: Request) -> Response:
    settings = _settings()
    signing_key = settings.anthropic_webhook_signing_key
    if not signing_key:
        return Response("ANTHROPIC_WEBHOOK_SIGNING_KEY is required", status_code=503)
    try:
        payload = await _read_limited_body(request, MAX_WEBHOOK_BODY_BYTES)
    except ValueError:
        return Response("request body too large", status_code=413)

    try:
        event = _webhook_client(settings).beta.webhooks.unwrap(
            payload,
            headers=dict(request.headers),
            key=signing_key,
        )
    except Exception:
        logger.exception("invalid webhook signature")
        return Response("invalid signature", status_code=401)

    if event.data.type == "session.status_run_started":
        start_queue_drain(settings)
        return Response(status_code=204)

    return Response(status_code=204)

File: examples/anthropic-managed-agents/javascript/src/app-webhook-server.ts
import Anthropic from "@anthropic-ai/sdk";
import { timingSafeEqual } from "node:crypto";
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";

import {
  DEFAULT_LOG_LEVEL,
  DEFAULT_SANDBOX_TIMEOUT_SECONDS,
  DEFAULT_TEMPLATE_NAME,
  DEFAULT_WORKER_MAX_IDLE_SECONDS,
} from "./constants.js";
import { JsonSandboxStore } from "./app-sandbox-store.js";
import { ensureWorkerSandbox } from "./sandbox-worker.js";
import { loadSettings, requireSetting, type Settings } from "./settings.js";

const store = new JsonSandboxStore();
const pendingWorkers = new Map<string, Promise<Awaited<ReturnType<typeof ensureWorkerSandbox>>>>();
const MAX_WEBHOOK_BODY_BYTES = 1_048_576;
const ROUTING_SCOPES = new Set(["session", "agent", "environment"]);

class PayloadTooLargeError extends Error {}

function webhookClient(settings: Settings) {
  return new Anthropic({
    apiKey: requireSetting(settings.anthropicApiKey, "ANTHROPIC_API_KEY"),
  });
}

function sessionId(event: unknown) {
  const id = (event as { data?: { id?: unknown } }).data?.id;
  if (!id) {
    throw new Error("webhook event does not include data.id");
  }
  return String(id);
}

function routingScope(settings: Settings) {
  const scope = settings.appSandboxRoutingScope ?? "session";
  if (!ROUTING_SCOPES.has(scope)) {
    throw new Error("APP_SANDBOX_ROUTING_SCOPE must be session, agent, or environment");
  }
  return scope;
}

async function routingTarget(settings: Settings, session: string) {
  const configuredEnvironmentId = requireSetting(
    settings.anthropicEnvironmentId,
    "ANTHROPIC_ENVIRONMENT_ID",
  );
  const scope = routingScope(settings);
  if (scope === "session") {
    return { environmentId: configuredEnvironmentId, routingScope: scope, routingId: session };
  }
  if (scope === "environment") {
    return {
      environmentId: configuredEnvironmentId,
      routingScope: scope,
      routingId: configuredEnvironmentId,
    };
  }

  const sessionInfo = await webhookClient(settings).beta.sessions.retrieve(session);
  if (sessionInfo.environment_id !== configuredEnvironmentId) {
    throw new Error(
      `session ${session} belongs to ${sessionInfo.environment_id}, but this worker is configured for ${configuredEnvironmentId}`,
    );
  }
  return {
    environmentId: sessionInfo.environment_id,
    routingScope: scope,
    routingId: sessionInfo.agent.id,
  };
}

async function ensureWorkerForEvent(settings: Settings, event: unknown) {
  const session = sessionId(event);
  const target = await routingTarget(settings, session);
  const key = `${target.routingScope}:${target.environmentId}:${target.routingId}`;
  const pending = pendingWorkers.get(key);
  if (pending) {
    return pending;
  }

  const worker = ensureWorkerForTarget(settings, target, session).finally(() => {
    pendingWorkers.delete(key);
  });
  pendingWorkers.set(key, worker);
  return worker;
}

async function ensureWorkerForTarget(
  settings: Settings,
  target: { environmentId: string; routingScope: string; routingId: string },
  session: string,
) {
  const assignment = await store.get(target);
  const sandbox = await ensureWorkerSandbox(settings, {
    sandboxId: assignment?.sandboxId,
    templateName: DEFAULT_TEMPLATE_NAME,
    timeoutSeconds: DEFAULT_SANDBOX_TIMEOUT_SECONDS,
    workerMaxIdleSeconds: DEFAULT_WORKER_MAX_IDLE_SECONDS,
    logLevel: DEFAULT_LOG_LEVEL,
  });
  await store.upsert({
    ...target,
    sessionId: session,
    sandboxId: sandbox.sandboxId,
  });
  return sandbox;
}

async function readBody(request: IncomingMessage, maxBytes: number) {
  const contentLength = request.headers["content-length"];
  if (contentLength && Number(contentLength) > maxBytes) {
    throw new PayloadTooLargeError("request body too large");
  }

  const chunks: Buffer[] = [];
  let size = 0;
  for await (const chunk of request) {
    const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
    size += buffer.length;
    if (size > maxBytes) {
      throw new PayloadTooLargeError("request body too large");
    }
    chunks.push(buffer);
  }
  return Buffer.concat(chunks).toString("utf8");
}

function writeJson(response: ServerResponse, statusCode: number, body: unknown) {
  response.writeHead(statusCode, { "content-type": "application/json" });
  response.end(JSON.stringify(body));
}

function startWorkerInBackground(settings: Settings, event: unknown) {
  void ensureWorkerForEvent(settings, event)
    .then((sandbox) => {
      console.log(`started worker sandbox ${sandbox.sandboxId}`);
    })
    .catch((error) => {
      console.error("failed to start worker sandbox", error);
    });
}

function hasAdminAccess(request: IncomingMessage, settings: Settings) {
  const expected = settings.appWebhookAdminToken;
  const authorization = request.headers.authorization;
  const actual = authorization?.startsWith("Bearer ") ? authorization.slice("Bearer ".length) : "";
  if (!expected || !actual) {
    return false;
  }

  const expectedBuffer = Buffer.from(expected);
  const actualBuffer = Buffer.from(actual);
  return expectedBuffer.length === actualBuffer.length && timingSafeEqual(expectedBuffer, actualBuffer);
}

async function handleWebhook(request: IncomingMessage, response: ServerResponse) {
  const settings = loadSettings();
  const signingKey = settings.anthropicWebhookSigningKey;
  if (!signingKey) {
    response.writeHead(503, { "content-type": "text/plain" });
    response.end("ANTHROPIC_WEBHOOK_SIGNING_KEY is required");
    return;
  }

  let body: string;
  try {
    body = await readBody(request, MAX_WEBHOOK_BODY_BYTES);
  } catch (error) {
    if (error instanceof PayloadTooLargeError) {
      response.writeHead(413, { "content-type": "text/plain" });
      response.end("request body too large");
      return;
    }
    throw error;
  }

  const client = webhookClient(settings);
  let event: ReturnType<typeof client.beta.webhooks.unwrap>;
  try {
    event = client.beta.webhooks.unwrap(body, {
      headers: Object.fromEntries(
        Object.entries(request.headers).map(([key, value]) => [
          key,
          Array.isArray(value) ? value.join(", ") : (value ?? ""),
        ]),
      ),
      key: signingKey,
    });
  } catch (error) {
    console.error(error);
    response.writeHead(401, { "content-type": "text/plain" });
    response.end("invalid signature");
    return;
  }

  if (event.data.type === "session.status_run_started") {
    startWorkerInBackground(settings, event);
    response.writeHead(204);
    response.end();
    return;
  }

  response.writeHead(204);
  response.end();
}

const server = createServer((request, response) => {
  if (request.method === "GET" && request.url === "/health") {
    writeJson(response, 200, { ok: true });
    return;
  }

  if (request.method === "GET" && request.url === "/sandboxes") {
    if (!hasAdminAccess(request, loadSettings())) {
      response.writeHead(401, { "content-type": "text/plain" });
      response.end("unauthorized");
      return;
    }

    void store
      .list()
      .then((assignments) => writeJson(response, 200, { sandboxes: assignments }))
      .catch((error) => {
        console.error(error);
        writeJson(response, 500, { error: "failed to read sandbox store" });
      });
    return;
  }

  if (request.method === "POST" && request.url === "/webhook") {
    void handleWebhook(request, response).catch((error) => {
      console.error(error);
      if (!response.headersSent) {
        response.writeHead(500, { "content-type": "text/plain" });
      }
      response.end("webhook handler failed");
    });
    return;
  }

  response.writeHead(404, { "content-type": "text/plain" });
  response.end("not found");
});

const port = Number(process.env.PORT ?? "8000");
server.listen(port, "0.0.0.0", () => {
  console.log(`app webhook server listening on ${port}`);
});
